### ISSUE TITLE
Consider a simpler build mode which does not do incremental builds

### ISSUE THREAD

**Issue by jakemac53** [dart-lang/build#3555]
We get a lot of complaints about build_runner not scaling to large projects. This is generally thought to be a result of how the analyzer is used in the package - basically any time a builder runs on a file it establishes a dependency on all files transitively imported by that file, which is O(N), and there is probably a fairly high constant multiplier as well. So you end up with O(N^2) complexity (where N is the # of libraries) just for the tracking of deps.

If we instead abandoned incrementality entirely, we wouldn't have this cost. You would regenerate all files every time you do a build, but I believe that overall it would be significantly faster, at least for large projects.

We could also stop computing hashes of files which does also take some time.

cc @natebosch wdyt? worth exploring or no?

**Comment by natebosch** [comment]
> just does the minimal work to ensure the necessary files are loaded.

It would still need to crawl transitive imports right?

> Then generate a package with some large number of files that all import each other, and run the script on it to see how it does?

I'm worried that indexing against a generated project could be misleading. I can imagine that different types of projects might have dependency graphs and library graphs with very different characteristics which might sway performance.

I wonder if we could get any analytics about real world asset graphs, file changes, and build performance.

**Comment by natebosch** [comment]
We could also consider drawing a harder distinction between local and pub packages, and set a boundary where we cache some work for pub packages, but use pessimistic invalidation over tracking for local code.

**Comment by jakemac53** [comment]
> It would still need to crawl transitive imports right?

Only once (possibly per phase) for the entire app.  Possibly could be optimized to add new sources early as well (as soon as they are written) and avoid doing it once per phase.

> I'm worried that indexing against a generated project could be misleading. I can imagine that different types of projects might have dependency graphs and library graphs with very different characteristics which might sway performance.

Yeah this would be a starting point, see how the worst case scenario stacks up. If it shows a significant improvement that is a sign its worth continuing to investigate (reach out to some people who have filed issues and have them try it on their real projects).

**Comment by jakemac53** [comment]
> We could also consider drawing a harder distinction between local and pub packages, and set a boundary where we cache some work for pub packages, but use pessimistic invalidation over tracking for local code.

Maybe, but I don't think this will help the common case. People run into this when their applications get large, and even if they split it up into multiple packages they are probably path dependencies usually.

**Comment by jakemac53** [comment]
We could optimize any package whose transitive deps had no codegen in the current build - but I think that still misses the use case of a single large application package, which I suspect is common practice.

**Comment by natebosch** [comment]
> Maybe, but I don't think this will help the common case.

The pessimistic invalidation over tracking is what would help this case - the distinction between local code and pub packages is trying to restore some of the caching we're otherwise giving up.

**Comment by jakemac53** [comment]
> The pessimistic invalidation over tracking is what would help this case - the distinction between local code and pub packages is trying to restore some of the caching we're otherwise giving up.

Yeah but I am pretty confident that the vast majority of code in these worst cases is app code, not pub package code.

**Comment by natebosch** [comment]
Would it be worthwhile to compute a separate digest which is based on transitive Dart code, as opposed to only computing the digest based on the asset content? I think we could compute it during the asset resolve, I'm not sure how hard it would be to feed it back to the build runner.

**Comment by jakemac53** [comment]
One idea could be to make a builder which only outputs these "transitive input" hashes next to each dart file. It would require all `.dart` inputs. When doing the transitive crawl in the analyzer, for each file we first check for that file, and if we can read then we immediately stop the crawl as long as the source is already loaded into the analysis driver (we now have a dependency on its contents).

The problem with this approach is that for sources in the same package this file wouldn't exist yet and the optimization wouldn't take effect. But we could get the optimization for dependencies I think using that approach?

**Comment by jakemac53** [comment]
The approach above could also allow optimization within a single package using `targets`.

**Comment by jakemac53** [comment]
We could also optimize this builder to use the "transitive input" hashes when present (which they would be for any dependencies of the "target").

**Comment by jakemac53** [comment]
I have a branch (pushed as a PR for easy viewing) which does attempt to do this https://github.com/dart-lang/build/pull/3556. I am not fully confident of its correctness yet and haven't analyzed the performance much.

I did quickly run it on the `build_modules` package which uses json_serializable and it was slower than the current build, at least initially.

I would love to get some initial community feedback on it though, you should be able to try it by just adding the following to your pubspec:

```
dependency_overrides:
  build_resolvers:
    git:
      url: https://github.com/dart-lang/build.git
      ref: transitive_deps
```

If you do try this out, please tell me how it works for you as well as your project structure. Do you have a single large app package or multiple sub-packages? Roughly how many files do you have per package?

**Comment by jakemac53** [comment]
I am going to close this issue - the worst case scenario builds should be significantly better now.
